tests/testthat/test_sim_data.R

```diff
@@ -0,0 +1,36 @@
+test_that("simulated data", {
+
+    library(dplyr)
+    library(serofoi)
+
+    seed <- 1234
+    sample_size_by_age <- 5
+    #----- Test for constant FoI
+    # foi_model <- "constant"
+    # foi_sim <- rep(0.02, 50)
+    # case_label <- "constant_foi_"
+    # max_lambda <- 0.035
+
+    #----- Test for stepwise decreasing FoI
+    foi_model = "tv_normal"
+    no_transm <- 0.0000000001
+    foi_sim <- c(rep(0.2, 25), rep(0.1, 10), rep(no_transm, 15))
+    case_label <- "sw_dec_foi_"
+    max_lambda <- 0.3
+
+    #----- Test function generate_sim_data
+    sim_data <- generate_sim_data(foi = foi_sim,
+                                  sample_size_by_age = sample_size_by_age,
+                                  tsur = 2050,
+                                  birth_year_min = 2000,
+                                  survey_label = 'foi_sim',
+                                  seed = seed)
```

Comment:
I know this function is stochastic but we can still test it works by varying the FOIs and checking that in limits it behaves as we'd like it to.

Reply:
Right now I'm relaying on model implementation to test this functionality. My idea is to make use of suitable models for different FOI trends. It is to be expected for a constant `foi_sim`to be correctly approximated by the `"constant"` model for all times, so what I do is to make sure that this lies in the confidence interval obtained by implementing this model:
```
# Define constant FoI for simulations
case_label <- "constant_foi_"
foi_model <- "constant"
foi_sim <- rep(0.02, tsur - birth_year_min)
max_lambda <- 0.035

# Generate simulated data and run the "constant" model
sim_data <- generate_sim_data(foi = foi_sim,
                              sample_size_by_age = sample_size_by_age,
                              tsur = tsur,
                              birth_year_min = birth_year_min,
                              survey_label = 'foi_sim',
                              seed = seed)
sim_seromodel <- run_seromodel(sim_data, foi_model = foi_model, n_iters = n_iters)

# Check consistency between sim_foi and the fitted foi
foi <- rstan::extract(sim_seromodel$seromodel_fit, "foi", inc_warmup = FALSE)[[1]]
foi_lower <- apply(foi, 2, function(x) quantile(x, 0.05))
foi_upper <- apply(foi, 2, function(x) quantile(x, 0.95))
expect_true(all((foi_sim >= foi_lower) & (foi_sim <= foi_upper)))
``` 
![plot_foi_const_test](https://github.com/epiverse-trace/serofoi/assets/45337127/64851b51-62ca-45ba-a838-76bd7bb4d9df)

I implemented this for a constant FOI and for the following smooth-decreasing FOI (red lines in the images):
```
case_label <- "smth_dec_foi_" #Smooth-decendent FoI
foi_model = "tv_normal"
foi_max = 0.2
stretch = 0.15
x <- 1:(tsur - birth_year_min)
foi_sim <- (-foi_max * (atan(stretch * (x - 25))) / (0.5 * pi) + foi_max) / 2
``` 
![plot_foi_smth_desc_test](https://github.com/epiverse-trace/serofoi/assets/45337127/234dd2fe-8b33-4055-be77-1f1be75477e2)

I think it's worth to open a separate issue to discuss this in more detail, in case that you're not convince by my approach.

Reply:
We should be able to test whether data we obtain from simulating from these models is as expected without resorting to solving the inverse problem (as this leaves us liable to a number of issues, e.g. the FOIs aren't identifiable given the data). We have analytical results for all of the models which should allow us to test these directly. E.g. for a constant FOI model then we will know approximately the probability that someone aged x is seropositive; if we use a large sample size, we can check that the simulated proportion is near to that value.